### ISSUE TITLE
On the fly publicPath when lazy-loading using @loadable/component?

### ISSUE THREAD

**Issue by andreasheim** [gregberge/loadable-components#661]
We have configured publicPath in ChunkExtractor, so it can be set dynamically/on the fly when deployed.  

This works for the initial bundles, publicPath set to `/pro/public` E.g. the paths look like this
```
https://local.moz.com/pro/public/bundle~._client_components_t.js
```

However it doesn't work for us when using lazy-loading via @loadable/component: 
```
const SelectReport = loadable(() => import('components/page-optimization/select-report'));
```

It will not use publicPath defined for ChunkExtractor: 
```
https://local.moz.com/pro/research/components-page-optimization-select-report~._client_components_page-optimization_se.js
```

If we define output.publicPath for webpack, this works in local dev. However we do rely on the 'on the fly' configuration to configure the paths for use with a CDN when deployed to prod. 

How do I make sure @loadable/component uses the publicPath defined for the chunk extractor? 

This is how we use ChunkExtractor: 
```
const getPublicPath = exports.getPublicPath = () => {
  if (yn(process.env.WEBPACK_DEV_MODE)) {
    return constants.SERVER.WEBPACK_PUBLIC_PATH;
  }

  const { ASSETS_PATH, CDN } = process.env;
  return CDN ? `${CDN}${ASSETS_PATH}` : ASSETS_PATH;
};

exports.getBundleElements = (bundle) => {
  const extractor = new ChunkExtractor({
    entrypoints: [bundle],
    publicPath: getPublicPath(),
    statsFile: STATS_FILE,
  });

  return {
    scripts: extractor.getScriptElements(),
    styles: extractor.getStyleElements(),
  };
};
```

I've been looking at the source code for @loadable/component, but I can't figure out how it resolves paths.

**Comment by andreasheim** [comment]
I tried to set `window.__webpack_public_path__ = '/pro/public'` but that didn't work either 
https://webpack.js.org/guides/public-path/

**Comment by theKashey** [comment]
It's just `__webpack_public_path__`, no window.

Loadable is not resolving any paths. 
- on the server side it uses `publicPath` provided during webpack built (publicPath), letting you override it.
- on the client it "uses" webpack, and webpack uses the current `publicPath` value. So just keep it in sync with the server.

**Comment by andreasheim** [comment]
I tried that, and it doesn't make a difference :/

(`window` is the default global namespace, so `window.__webpack_public_path__` and `__webpack_public_path__` should be the same)

**Comment by andreasheim** [comment]
Oh - it needs to be in the entry script, and without `window` like you said. 

```
__webpack_public_path__ = '/pro/public/';
```

as the first line in my entry script does the trick. Thank you!

**Comment by andreasheim** [comment]
May I suggest adding this to the documentation? I understand this may be more of a webpack "thing" than loadable, but seems to be an easy stumbling block. 

Love using the library so far 😃

**Comment by theKashey** [comment]
So yeah `__webpack_public_path__` has to be inside webpack bundle to control that bundle. However, you can do
`__webpack_public_path__ = window.__webpack_public_path__;` and webpack will understand which variable is your global variable, and which one is it's configuration one.

If you have an idea about where this moment has to appear in the documentation - feel free to contrinute.

**Comment by andreasheim** [comment]
I think it should go into the api documentation for @loadable/component - https://loadable-components.com/docs/api-loadable-component/ 

Something like a note before `options.resolveComponent`.

If you are defining `publicPath` for ChunkExtractor, e.g. to prefix a CDN in production, you will likely need to use webpack's ['on the fly' publicPath](https://webpack.js.org/guides/public-path/#on-the-fly) configuration in your bundles. In your entry script, set the special global `__webpack_public_path__` variable, and match the path set for ChunkExtractor. 
```
__webpack_public_path__ = 'your_public_path_goes_here';
```

**Comment by theKashey** [comment]
Usually, that's working in the opposite way - first, you discover the need to override `publicPath`, then you are discovering extra options to `ChunkExtractor`.

So here is the question - why you did it the other way? Because you were playing from SSR, not CSR? (Majority don't have SSR in the dev mode and first run into the webpack part of a problem, and only then into the SSR-ed paths

**Comment by andreasheim** [comment]
Our app had been using SSR-light - using a library to render the html with react. With that setup we couldn't figure out how to have webpack inject the bundles, so we had injected them manually. We had left the splitting setting at the defaults, which creates a small number of bundles, so it was manageable at first. This also meant we were able to control the public path. 

However we we wanted to automate that, and also split the bundles further. I already knew we wanted to use @loadable for lazy-loading, and so we started with using ChunkExtractor to set things up properly server-side first.
